### DISCUSSION TITLE
OnBeforePopup creates and sizes window even if one is supplied in newBrowser

### DISCUSSION THREAD

**Discussion by aimcosoftware** [cefsharp/CefSharp#4438] [General]
Trying to use onBeforePopup() to create and size a ChromiumWebBrowser as a tab.
Returning the created browser in newBrowser is ignored, a new window is created anyway.

The only way to do that seems to be to create and navigate my own browser, and cancel the default.
This means window.open() doesn't have a reference to the created window.

Anything that can be done to change this behaviour?

**Comment by amaitland** [comment]
> Trying to use onBeforePopup() to create and size a ChromiumWebBrowser as a tab.
> Returning the created browser in newBrowser is ignored, a new window is created anyway.

Are you using `WinForms` or `WPF`? What does your code look like?

If you are using `WinForms` you can use `ChromiumHostControl` as per https://github.com/cefsharp/CefSharp/pull/3529#issuecomment-828934239 This isn't quite as advanced as using `ChromiumWebBrowser`, it is better tested and more examples.

**Reply by aimcosoftware** [reply]
It's Winforms, I have moved on a little, in that I am now using SetAsChild() to prevent creation of a new window.
I have a wrapper around the browser to add functionality. The most basic code I have tried is this:

```
Function OnBeforePopup(WebBrowser As IWebBrowser, browser As IBrowser, frame As IFrame, targetUrl As String, targetFrameName As String, targetDisposition As WindowOpenDisposition, userGesture As Boolean, popupFeatures As IPopupFeatures, windowInfo As IWindowInfo, browserSettings As IBrowserSettings, ByRef noJavascriptAccess As Boolean, ByRef newBrowser As IWebBrowser) As Boolean Implements ILifeSpanHandler.OnBeforePopup

  'Create wrapper and set underlying ChromiumWebBrowser to newBrowser
  Dim Wrapper = New BrowserWrapper()
  TabForm.Controls.Add(Wrapper)
  windowInfo.SetAsChild(Wrapper.Browser)
  newBrowser = Wrapper.Browser
  Return newBrowser Is Nothing
End Function

```
After that, it seems that the window handles get corrupted as it does odd things, like clicking a link in the opener, opens the link in the popup.

Rather than waste your time, I'll keep investigating, as I am making some progress.

**Reply by amaitland** [reply]
If possible I'd suggest using a `ChromiumHostControl`. It's much better tested, has examples. 

If you really need to use `ChromiumWebBrowser` then it's quite a bit more complex. I haven't tried this approach in many years, there were `upstream CEF` issues, so I removed all examples. The last example exists in the `M57` branch, so it's quite possible the `CEF` issues have been resolved since then. From memory one of the biggest issues was that `PostData` was propagated to popups incorrectly.

The old example is at https://github.com/cefsharp/CefSharp/blob/cefsharp/57/CefSharp.WinForms.Example/Handlers/LifeSpanHandler.cs#L19 (It's commented out)

The [LifeSpanHandler](https://github.com/cefsharp/CefSharp/blob/cefsharp/111/CefSharp.WinForms/Handler/LifeSpanHandler.cs) implementation used for `ChromiumHostControl` is much more advanced than any of the old examples ever were. You can in theory use it as a reference and rewrite your own implementation to use `ChromiumWebBrowser`.

**Reply by aimcosoftware** [reply]
OK thanks. I probably need to use ChromiumWebBrowser, as the project is to large to change it now, but I will look at the old example and the ChromiumHostControl code for reference.

**Reply by amaitland** [reply]
> I probably need to use ChromiumWebBrowser, as the project is to large to change it now, but I will look at the old example and the ChromiumHostControl code for reference.

For what it's worth both inherit from `ChromiumHostControlBase` which provides some of the common functionality.

**Reply by aimcosoftware** [reply]
Sorry to trouble you, it was a basic error in my code:
`
windowInfo.SetAsChild(Wrapper.Browser.Handle)
`
Should be:
`
windowInfo.SetAsChild(WebBrowser.ParentWrapper.Browser.Handle)
`
Where `WebBrowser.ParentWrapper` is an extension to get my Wrapper.

Thanks again.

**Reply by amaitland** [reply]
> windowInfo.SetAsChild(WebBrowser.ParentWrapper.Browser.Handle)

Is `WebBrowser.ParentWrapper.Browser.Handle` actually `ChromiumWebBrowser.Handle`? It's really hard for me to tell with the abstraction layer. For resizes to work properly you'll need to make sure the correct handle is used as the parent.

**Reply by aimcosoftware** [reply]
Sorry, yes it is. Resizing does work correctly, and no more odd window behaviour.
Only other thing I need to discover is when a popup with a targetFrameName closes, CEF still holds a reference, so doesn't raise onBeforePopup() again.

**Reply by amaitland** [reply]
> Only other thing I need to discover is when a popup with a targetFrameName closes,

[IFrame.Name](https://cefsharp.github.io/api/110.0.x/html/P_CefSharp_IFrame_Name.htm) for the `MainFrame` should match the `targetFrameName` if one was specified.

The following should get you the `Frame Name`.
```c#
var name = browser.MainFrame.Name;
```

browser is an instance of `IBrowser`.

**Reply by aimcosoftware** [reply]
Thanks for the pointers, still got some odd behaviour but to get this working, I need to learn some more.
There are still some window handle issues, when using `WindowInfo.SetAsChild()`. I haven't investigated fully yet.
1) If CEF navigates newBrowser it's window isn't visible, but it is when the popup is navigated before setting newBrowser.
2) Calling `ShowDevTools()` opens a window, but navigates the popup to DevTools.
As you say, maybe these are things that have never been fixed, or there is something I am missing.

Creating my own popup and cancelling OnBeforePopup, works exactly as expected, but that leaves the original problem of the the popup not being parented inside CEF.

**Reply by aimcosoftware** [reply]
@amaitland I don't think SetAsChild() is what I need, but can't find any info on what I do need.
Can you offer anything?
I just need to is to pass my ChromiumWebBrowser back to CEF for it to use e.g:
```
Dim Wrapper As New MyWrapper()
Wrapper.CreateControl()

Wrapper.Browser = New ChromiumWebBrowser()
Wrapper.Browser.BrowserSettings = GlobalSettings()
Wrapper.Controls.Add(Browser)

'Pass back Wrapper.Browser to be navigated by CEF.
newBrowser = Wrapper.Browser
???
```

**Reply by aimcosoftware** [reply]
Followed the source from Winforms.Handler.LifespanHandler, so it's now working.
There are still some issues trying to use ShowDevTools, so need to work on that.

**Reply by aimcosoftware** [reply]
After successfully creating a new tab/popup with `window.open()` , then calling `ShowDevTools()`, opens a popup copy of the current browser, instead of DevTools. Any idea why that is?

**Reply by amaitland** [reply]
> After successfully creating a new tab/popup with `window.open()` , then calling `ShowDevTools()`

Are you calling `ShowDevTools` on the popup or the main browser?

> , opens a popup copy of the current browser, instead of DevTools. Any idea why that is?

What does your `LifeSpanHandler` code look like?

**Reply by aimcosoftware** [reply]
Calling ShowDevTools on the browserHost of the popup. Calling it on the opener browser works OK.

```
Function DoClose(WebBrowser As IWebBrowser, browser As IBrowser) As Boolean Implements ILifeSpanHandler.DoClose
	If browser.MainFrame.Url.StartsWith("devtools") Then
		Return False
	Else
		Dim Web = WebBrowser.GetDOMView
		Web.Invoke(Sub() Web.Close())
		Return True
	End If
End Function

Sub OnAfterCreated(WebBrowser As IWebBrowser, browser As IBrowser) Implements ILifeSpanHandler.OnAfterCreated
	If WebBrowser.GetDOMBase.DevToolsHandle = -1 Then
		WebBrowser.GetDOMBase.DevToolsHandle = browser.GetHost.GetWindowHandle
	End If
End Sub

Sub OnBeforeClose(WebBrowser As IWebBrowser, browser As IBrowser) Implements ILifeSpanHandler.OnBeforeClose
	If browser.MainFrame.Url.StartsWith("devtools") Then
		WebBrowser.GetDOMView.DevToolsHandle = 0
		SaveDevToolsLocation(browser.GetHost)
	End If
End Sub

Function OnBeforePopup(WebBrowser As IWebBrowser, browser As IBrowser, frame As IFrame, targetUrl As String, targetFrameName As String, targetDisposition As WindowOpenDisposition, userGesture As Boolean, popupFeatures As IPopupFeatures, windowInfo As IWindowInfo, browserSettings As IBrowserSettings, ByRef noJavascriptAccess As Boolean, ByRef newBrowser As IWebBrowser) As Boolean Implements ILifeSpanHandler.OnBeforePopup
	Dim Web = WebBrowser.GetDOMView
	Dim Evt As NewWindowRequestedArgs
	Evt.Browser = Nothing
	Web.Invoke(
		Sub()
			Evt.Target = targetFrameName
			Evt.Url = targetUrl
			Evt.Type = targetDisposition
			If Evt.Type = NewWindowType.NewPopup Or Evt.Type = NewWindowType.NewWindow Then
				Evt.Left = Val(popupFeatures.X)
				Evt.Top = Val(popupFeatures.Y)
				Evt.Width = Val(popupFeatures.Width)
				Evt.Height = Val(popupFeatures.Height)
			End If

			'Get new browser
			Web.RaiseNewWindowEvent(Evt)

			If Evt.Browser IsNot Nothing Then
				'Create ChromiumWebBrowser add it to Controls
				Evt.Browser.Initialize()
				'Evt.Browser.Browser is ChromiumWebBrowser
				windowInfo.SetAsChild(Evt.Browser.Browser.Handle) 
			End If
		End Sub)
	newBrowser = Evt.Browser.Browser
	Return newBrowser Is Nothing
End Function
```

**Reply by aimcosoftware** [reply]
This is what I am seeing:
Open Google Tab, ShowDevTools - OK
Call window.Open() from Google DevTools, page opens in tab - OK
Call ShowDevTools from 2nd tab, popup opens with DevTools as title, but calling page as content.
Closing the 2nd DevTools window, calls DoClose()

![image](https://user-images.githubusercontent.com/100280982/228657345-b9f978b3-90fb-4143-8cb5-1a3cb60ce95f.png)

**Reply by aimcosoftware** [reply]
@amaitland  Just FYI: I've removed the call to WindowInfo.SetAsChild(), tried a few other changes, and it's the same. Seems to be related to passing back newBrowser.

**Reply by amaitland** [reply]
Confirming I can reproduce the problem, unfortunately it looks like a bug in `CEF`.

If we pass `nullptr` into `ShowDevTools` instead of the `CefClient` instance then it shows the popup as expected. Unfortunately this would mean no handlers would be called for `DevTools`. 

https://github.com/cefsharp/CefSharp/blob/cefsharp/111/CefSharp.Core.Runtime/Internals/CefBrowserHostWrapper.cpp#L186

Potentially we could conditionally pass in `nullptr`.

**Reply by aimcosoftware** [reply]
OK, thanks very much for your help and confirming, I'll report it in CEF.

**Reply by aimcosoftware** [reply]
@amaitland Just FYI, there are also issues if you just open the default popup.
Called `window.open()` (executed script, not from DevTools) and returned false in `OnBeforePopup`.
```
Function OnBeforePopup(WebBrowser As IWebBrowser, browser As IBrowser, frame As IFrame, targetUrl As String, targetFrameName As String, targetDisposition As WindowOpenDisposition, userGesture As Boolean, popupFeatures As IPopupFeatures, windowInfo As IWindowInfo, browserSettings As IBrowserSettings, ByRef noJavascriptAccess As Boolean, ByRef newBrowser As IWebBrowser) As Boolean Implements ILifeSpanHandler.OnBeforePopup
  Return False
End Function
```
The opener window now gets events for the popup. The icon on the Google Tab, is from the Packing Partner popup, but messages stop because it gets `FrameLoadStart`, but not `FrameLoadEnd`.
![image](https://user-images.githubusercontent.com/100280982/228804427-3aa8df3d-8fcb-4bb3-b999-4364c01c4ecb.png)

**Reply by aimcosoftware** [reply]
Trying to think of a workaround, so that `OnBeforePopup` can return true, and inject it into script, e.g:
```
OldOpen = open

function open(a,b,c){
  OldOpen(a,b,c)
  return getPopupRefFromCefSharp()
}
```
Do you have any ideas?

**Reply by amaitland** [reply]
> The opener window now gets events for the popup. The icon on the Google Tab, is from the Packing Partner popup, but messages stop because it gets `FrameLoadStart`, but not `FrameLoadEnd`.

The parent `ChromiumWebBrowser` should only fire `FrameLoadStart` and `FrameLoadEnd` events for itself (`browser.IsPopup == false`). Have you implemented `LoadHandler` to get the popup related calls? Using the following https://jsfiddle.net/nkdvpcma/ I'm seeing the methods called for the popup.

```c#
browser.LoadHandler = new LoadHandlerEx();

private class LoadHandlerEx : CefSharp.Handler.LoadHandler
{
	protected override void OnFrameLoadStart(IWebBrowser chromiumWebBrowser, FrameLoadStartEventArgs frameLoadStartArgs)
	{
		base.OnFrameLoadStart(chromiumWebBrowser, frameLoadStartArgs);
	}

	protected override void OnFrameLoadEnd(IWebBrowser chromiumWebBrowser, FrameLoadEndEventArgs frameLoadEndArgs)
	{
		base.OnFrameLoadEnd(chromiumWebBrowser, frameLoadEndArgs);
	}
}
```

> Trying to think of a workaround, so that `OnBeforePopup` can return true, and inject it into script, e.g:

I've no idea how that would work exactly.

**Reply by amaitland** [reply]
> unfortunately it looks like a bug in `CEF`.

After reviewing the `CEF` source code in more detail I couldn't see anything that would explain the behaviour. Going with the assumption the behaviour originates from `CefSharp` I spent a couple of hours debugging. I can now confirm the problem is actually in `CefSharp`. 

Basically the `IWebBrowserInternal.Set*` methods are being called for the `DevTools` popup, the call to `OnAfterBrowserCreated` ends up with a call to `LoadUrl` which overrides the `DevTools popup`.

**Reply by aimcosoftware** [reply]
I'm not add a specific LoadHandler, just using the FrameLoadxxxx events in ChromiumWebBrowser.

**Reply by aimcosoftware** [reply]
> I can now confirm the problem is actually in `CefSharp`

Ah OK, thanks for your time on this. I'm still a bit confused by the default popup events going to the opener, I need to look closer at that.

**Reply by amaitland** [reply]
PR #4443 should resolve this, I've marked it as draft as it will still need some testing.

**Reply by aimcosoftware** [reply]
Great thanks, will give it a try when it's released.

**Reply by aimcosoftware** [reply]
@amaitland  Thanks, can confirm that DevTools opens as expected, but there are still other issues related to events being raised on the wrong browser. I'll start a new discussion for that.
